### PR TITLE
Add locations to all types

### PR DESCRIPTION
**Pull request by mlutze**
Part of https://github.com/flix/flix/issues/2120

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala [hunk 1]
```diff
@@ -450,16 +450,16 @@ object Typer extends Phase[KindedAst.Root, TypedAst.Root] {
...
           (constrs, bodyType, bodyEff) <- visitExp(exp)
           _ <- unifyTypeM(argType, argTypeVar, loc)
-          resultTyp <- unifyTypeM(tvar, Type.mkArrowWithEffect(argType, bodyEff, bodyType), loc)
+          resultTyp <- unifyTypeM(tvar, Type.mkArrowWithEffect(argType, bodyEff, bodyType, loc), loc)
         } yield (constrs, resultTyp, Type.Pure)
 
       case KindedAst.Expression.Apply(exp, exps, tvar, evar, loc) =>
-        val lambdaBodyType = Type.freshVar(Kind.Star)
-        val lambdaBodyEff = Type.freshVar(Kind.Bool)
+        val lambdaBodyType = Type.freshVar(Kind.Star, loc = loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Typer.scala:457]: Whats the reason for the use of named parameters suddenly? I don't think `loc = loc` adds much value.

File: main/src/ca/uwaterloo/flix/language/phase/Kinder.scala [hunk 2]
```diff
@@ -219,18 +219,18 @@ object Kinder extends Phase[ResolvedAst.Root, KindedAst.Root] {
     */
   private def visitExp(exp00: ResolvedAst.Expression, kenv: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Expression, KindError] = exp00 match {
 
-    case ResolvedAst.Expression.Wild(loc) => KindedAst.Expression.Wild(Type.freshVar(Kind.Star), loc).toSuccess
+    case ResolvedAst.Expression.Wild(loc) => KindedAst.Expression.Wild(Type.freshVar(Kind.Star, loc = loc), loc).toSuccess
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Kinder.scala:222]: Also here with the `loc = `loc`

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala (hunk 1 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Typer.scala:457]: Type.freshVar has optional parameters. `loc` is not optional but it comes after the optional ones so the named param style is required. I'll move it ahead of the optional params.

File: main/src/ca/uwaterloo/flix/language/ast/Type.scala [hunk 3]
```diff
@@ -797,12 +839,12 @@ object Type {
     *
     * Must not be used before kinding.
     */
-  def mkOr(tpe1: Type, tpe2: Type): Type = (tpe1, tpe2) match {
+  def mkOr(tpe1: Type, tpe2: Type, loc: SourceLocation = SourceLocation.Unknown): Type = (tpe1, tpe2) match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Type.scala:842]: I am not so happy that we allow a default parameters with `Unknown`.

The fact that we have `Unknown` locations is really a design mistake it Flix- they suddenly pop up in error messages and then there is total confusion. 

Are there at least some of these we could make non-default and then we would have to explicitly pass in `Unknown` -- but only in a few places? If its hundreds of places we probably have to live with the default param.

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Type.scala]: Commit aadba6354ae91050827938a744c3f5d7c45c1e83 removes these. There were about a hundred of them at first until Intellij found more...

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 4]
```diff
@@ -270,14 +270,14 @@ object Simplifier extends Phase[TypedAst.Root, SimplifiedAst.Root] {
       case TypedAst.Expression.Spawn(exp, tpe, eff, loc) =>
         // Wrap the expression in a closure: () -> tpe & eff
         val e = visitExp(exp)
-        val lambdaTyp = Type.mkArrowWithEffect(Type.Unit, eff, e.tpe)
+        val lambdaTyp = Type.mkArrowWithEffect(Type.Unit, eff, e.tpe, SourceLocation.Unknown)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:273]: Here we could just use `loc`

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 5]
```diff
@@ -270,14 +270,14 @@ object Simplifier extends Phase[TypedAst.Root, SimplifiedAst.Root] {
...
         val lambdaExp = SimplifiedAst.Expression.Lambda(List(), e, lambdaTyp, loc)
         SimplifiedAst.Expression.Spawn(lambdaExp, lambdaTyp, loc)
 
       case TypedAst.Expression.Lazy(exp, tpe, loc) =>
         // Wrap the expression in a closure: () -> tpe & Pure
         val e = visitExp(exp)
-        val lambdaTyp = Type.mkArrowWithEffect(Type.Unit, Type.Pure, e.tpe)
+        val lambdaTyp = Type.mkArrowWithEffect(Type.Unit, Type.Pure, e.tpe, SourceLocation.Unknown)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:280]: same here.

File: main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala [hunk 6]
```diff
@@ -144,7 +144,7 @@ object Unification {
         } else {
           // Case 1.2: The fields do not match, attempt to match with a field further down.
           visit(restRow2) map {
-            case (subst, rewrittenRow) => (subst, Type.mkRecordExtend(field2, fieldType2, rewrittenRow))
+            case (subst, rewrittenRow) => (subst, Type.mkRecordExtend(field2, fieldType2, rewrittenRow, SourceLocation.Unknown))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala:147]: We should have access to a lot of source locs here! (they are just underscored...)

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala [hunk 7]
```diff
@@ -533,7 +533,7 @@ object Typer extends Phase[KindedAst.Root, TypedAst.Root] {
             (constrs1, tpe1, eff1) <- visitExp(exp1)
             (constrs2, tpe2, eff2) <- visitExp(exp2)
             resultType <- unifyTypeM(tvar, tpe1, tpe2, Type.Bool, loc)
-            resultEff = Type.mkAnd(eff1, eff2)
+            resultEff = Type.mkAnd(eff1, eff2, SourceLocation.Unknown)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Typer.scala:536]: We have `loc`.

**Review by magnus-madsen — COMMENTED** [review]
Looks good, but we can do a lot more to use local `locs` instead of just using Unknown everywhere.

At one point I had uses of `Unknown` down to less than 10? I think uses!

Lets not make it into 50 extra with this PR

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 8]
```diff
@@ -70,21 +70,21 @@ object Symbol {
...
-    new VarSym(flix.genSym.freshId(), ident.name, Type.freshUnkindedVar(), Scopedness.Unscoped, ident.loc)
+    new VarSym(flix.genSym.freshId(), ident.name, Type.freshUnkindedVar(loc = ident.loc), Scopedness.Unscoped, ident.loc)
   }
 
   /**
     * Returns a fresh variable symbol for the given identifier and scopedness.
     */
   def freshVarSym(ident: Name.Ident, scopedness: Scopedness)(implicit flix: Flix): VarSym = {
-    new VarSym(flix.genSym.freshId(), ident.name, Type.freshUnkindedVar(), scopedness, ident.loc)
+    new VarSym(flix.genSym.freshId(), ident.name, Type.freshUnkindedVar(loc = ident.loc), scopedness, ident.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:80]: unnecessary labelled arg?

File: main/src/ca/uwaterloo/flix/language/ast/Symbol.scala [hunk 9]
```diff
@@ -70,21 +70,21 @@ object Symbol {
     * Returns a fresh variable symbol for the given identifier.
     */
   def freshVarSym(ident: Name.Ident)(implicit flix: Flix): VarSym = {
-    new VarSym(flix.genSym.freshId(), ident.name, Type.freshUnkindedVar(), Scopedness.Unscoped, ident.loc)
+    new VarSym(flix.genSym.freshId(), ident.name, Type.freshUnkindedVar(loc = ident.loc), Scopedness.Unscoped, ident.loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Symbol.scala:73]: unnecessary labelled arg?

File: main/src/ca/uwaterloo/flix/language/ast/ops/TypedAstOps.scala [hunk 10]
```diff
@@ -4,7 +4,7 @@ import ca.uwaterloo.flix.language.ast.Ast.Annotation.{Benchmark, Law, Lint, Test
 import ca.uwaterloo.flix.language.ast.Ast.HoleContext
 import ca.uwaterloo.flix.language.ast.TypedAst.Predicate.{Body, Head}
 import ca.uwaterloo.flix.language.ast.TypedAst._
-import ca.uwaterloo.flix.language.ast.{Symbol, Type}
+import ca.uwaterloo.flix.language.ast.{SourceLocation, Symbol, Type}
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ops/TypedAstOps.scala:7]: unused import?

File: main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala [hunk 11]
```diff
@@ -19,7 +19,7 @@ package ca.uwaterloo.flix.language.phase
 import ca.uwaterloo.flix.api.Flix
 import ca.uwaterloo.flix.language.CompilationError
 import ca.uwaterloo.flix.language.ast.SimplifiedAst._
-import ca.uwaterloo.flix.language.ast.{Ast, Symbol, Type}
+import ca.uwaterloo.flix.language.ast.{Ast, SourceLocation, Symbol, Type}
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ClosureConv.scala:22]: unused import?

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 12]
```diff
@@ -1019,7 +1019,7 @@ object Lowering extends Phase[Root, Root] {
     */
   private def box(exp: Expression)(implicit root: Root, flix: Flix): Expression = {
     val loc = exp.loc
-    val tpe = Type.mkPureArrow(exp.tpe, Types.Boxed)
+    val tpe = Type.mkPureArrow(exp.tpe, Types.Boxed, SourceLocation.Unknown)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1022]: We have e.g. `exp.loc`

File: main/src/ca/uwaterloo/flix/language/phase/Lowering.scala [hunk 13]
```diff
@@ -1200,7 +1200,7 @@ object Lowering extends Phase[Root, Root] {
     * Returns a pure tuple expression constructed from the given list of expressions `exps`.
     */
   private def mkTuple(exps: List[Expression], loc: SourceLocation): Expression = {
-    val tpe = Type.mkTuple(exps.map(_.tpe))
+    val tpe = Type.mkTuple(exps.map(_.tpe), loc)
     val eff = Type.Pure
     Expression.Tuple(exps, tpe, eff, loc)
   }
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:1206]: This Phase contains _alot_ of `Unknown`s that can be avoided.

File: main/src/ca/uwaterloo/flix/language/phase/unification/BoolUnification.scala [hunk 14]
```diff
@@ -330,29 +330,29 @@ object BoolUnification {
       //                println(s.substring(0, Math.min(len, 300)))
       //              }
 
-      Type.Apply(Type.Apply(Type.Or, tpe1), tpe2)
+      Type.Apply(Type.Apply(Type.Or, tpe1, SourceLocation.Unknown), tpe2, SourceLocation.Unknown)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/BoolUnification.scala:333]: Could we not just use tpe1.loc or tpe2.loc?

**Review by magnus-madsen — COMMENTED** [review]
I am sending this back for one more round because I think some labelled args and uses of Unknown can be eliminated.

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**